### PR TITLE
query(join): improve join results gather[7X]

### PR DESCRIPTION
**Pull request by sundy-li**
I hereby agree to the terms of the CLA available at: https://databend.rs/dev/policies/cla/

## Summary

Summary about this PR

```
databend > select  avg(a.number) from numbers(100000000) as a inner join numbers(1000000) b on a.number = b.number + 1;
+---------------+
| avg(a.number) |
+---------------+
|      500000.5 |
+---------------+
1 row in set (3.842 sec)
```

After

```
databend > select  avg(a.number) from numbers(100000000) as a inner join numbers(1000000) b on a.number = b.number + 1;
+---------------+
| avg(a.number) |
+---------------+
|      500000.5 |
+---------------+
1 row in set (0.526 sec)

```

## Changelog

 
- Improvement
 

## Related Issues

Fixes #issue

### CODE REVIEW

**Review by xudong963 — COMMENTED** [review]
Cool!

**Comment by xudong963** [comment]
@mergify update

**Comment by xudong963** [comment]
Hi @sundy-li , two tpch tests failed 
image

**Comment by sundy-li** [comment]
> Hi @sundy-li , friendly reminder: two tpch tests failed image

Thanks, I'm working on resolving them.

**Comment by sundy-li** [comment]
There is an infinite loop in the subquery executor to execute `13_0021_q21.sql`, I still did not find the reason.  cc @leiysky  may need your help.
```
2022-06-25T14:28:55.853175Z  WARN databend_query::pipelines::new::executor::pipeline_pulling_executor: Send finish event error "Disconnected(..)"
2022-06-25T14:28:55.873673Z  WARN databend_query::pipelines::new::executor::pipeline_pulling_executor: Send finish event error "Disconnected(..)"
2022-06-25T14:28:55.892633Z  WARN databend_query::pipelines::new::executor::pipeline_pulling_executor: Send finish event error "Disconnected(..)"
2022-06-25T14:28:55.910599Z  WARN databend_query::pipelines::new::executor::pipeline_pulling_executor: Send finish event error "Disconnected(..)"
```

**Comment by xudong963** [comment]
> There is an infinite loop in the subquery executor to execute `13_0021_q21.sql`, I still did not find the reason. cc @leiysky may need your help.
> 
> ```
> 2022-06-25T14:28:55.853175Z  WARN databend_query::pipelines::new::executor::pipeline_pulling_executor: Send finish event error "Disconnected(..)"
> 2022-06-25T14:28:55.873673Z  WARN databend_query::pipelines::new::executor::pipeline_pulling_executor: Send finish event error "Disconnected(..)"
> 2022-06-25T14:28:55.892633Z  WARN databend_query::pipelines::new::executor::pipeline_pulling_executor: Send finish event error "Disconnected(..)"
> 2022-06-25T14:28:55.910599Z  WARN databend_query::pipelines::new::executor::pipeline_pulling_executor: Send finish event error "Disconnected(..)"
> ```

Oh, how large dataset do you use to run q21? 
The messages are from here: https://github.com/datafuselabs/databend/blob/main/query/src/pipelines/new/processors/transforms/transform_apply.rs#L252. I think it isn't a dead loop, but the result has not been run yet 🤣

**Comment by leiysky** [comment]
> There is an infinite loop in the subquery executor to execute `13_0021_q21.sql`, I still did not find the reason.

I've just opened a PR, which can probably fix the error. https://github.com/datafuselabs/databend/pull/6232

**Comment by sundy-li** [comment]
> I've just opened a PR, which can probably fix the error. #6232

Works after merge main branch.  But now `semi`, `anti`, `left` join are not optimized, will be done in future pr.

File: query/src/pipelines/new/processors/transforms/hash_join/row.rs [hunk 1]
```diff
@@ -86,40 +86,32 @@ impl RowSpace {
...
+
+        for row_ptr in row_ptrs.iter() {
+            indices.push((
+                row_ptr.chunk_index as usize,
+                row_ptr.row_index as usize,
+                1usize,
+            ));
         }
 
         // If build_key doesn't have duplicated columns, the length of row_ptrs will be one.
```
Comment (xudong963) [inline query/src/pipelines/new/processors/transforms/hash_join/row.rs:106]: The annotation is stale

File: query/src/pipelines/new/processors/transforms/hash_join/chaining_hash_table.rs [hunk 2]
```diff
@@ -392,6 +435,23 @@ impl ChainingHashTable {
         Ok(result_block)
     }
 
+    // Merge build block and probe block
```
Comment (xudong963) [inline query/src/pipelines/new/processors/transforms/hash_join/chaining_hash_table.rs:438]:
```suggestion
    // Merge build block and probe block that have the same number of rows 
```

**Review by xudong963 — APPROVED** [review]
Rest LGTM

**Review by leiysky — APPROVED** [review]

**Pull request merged**